packages/miniflare/src/workers/d1/dumpSql.ts

```diff
@@ -0,0 +1,152 @@
+// NOTE: this function duplicates the logic inside SQLite's shell.c.in as close
+// as possible, with any deviations noted.
+import { SqlStorage } from "@cloudflare/workers-types/experimental";
+
+export function* dumpSql(
+	db: SqlStorage,
+	options?: {
+		noSchema?: boolean;
+		noData?: boolean;
+		tables?: string[];
+	}
+) {
+	yield `PRAGMA defer_foreign_keys=TRUE;`;
+
+	// Empty set means include all tables
+	const filterTables = new Set(options?.tables || []);
+	const { noData, noSchema } = options || {};
+
+	// Taken from SQLite shell.c.in https://github.com/sqlite/sqlite/blob/master/src/shell.c.in#L8444-L8450
+	// @ts-ignore (SqlStorageStatement needs to be callable)
+	const tables_cursor = db.prepare(`
+    SELECT name, type, sql 
+      FROM sqlite_schema AS o 
+    WHERE (true) AND type=='table' 
+      AND sql NOT NULL 
+    ORDER BY tbl_name='sqlite_sequence', rowid;
+  `)();
+	const tables: any[] = Array.from(tables_cursor);
+
+	for (const { name: table, sql } of tables) {
+		if (filterTables.size > 0 && !filterTables.has(table)) continue;
+
+		if (table === "sqlite_sequence") {
+			if (!noSchema) yield `DELETE FROM sqlite_sequence;`;
+		} else if (table.startsWith(`sqlite_stat`)) {
```

Comment:
This matches too many table names!  It's supposed to only match sqlite_stat plus exactly one more character.   The code in shell.in.c's `dump_callback` is `sqlite3_strglob("sqlite_stat?", zTable)==0`, and the `?` operator in a glob matches exactly one character.  I don't know if JS has globbing, but the equivalent regexp would be `/sqlite_stat./`.

Reply:
Good catch! Fixed.